// jess/lib/tree/list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.list = exports.List = void 0;
const node_1 = require("./node");
/**
 * A list of expressions
 *
 * i.e. one, two, three
 * or .sel, #id.class, [attr]
 */
class List extends node_1.Node {
    toArray() {
        return this.value;
    }
    toCSS(context, out) {
        out.add('', this.location);
        const length = this.value.length - 1;
        const pre = context.pre;
        const cast = context.cast;
        this.value.forEach((node, i) => {
            const val = cast(node);
            val.toCSS(context, out);
            if (i < length) {
                if (context.inSelector) {
                    out.add(`,\n${pre}`);
                }
                else {
                    out.add(', ');
                }
            }
        });
    }
    toModule(context, out) {
        out.add(`$J.list([\n`, this.location);
        context.indent++;
        let pre = context.pre;
        const length = this.value.length - 1;
        this.value.forEach((node, i) => {
            out.add(pre);
            if (node instanceof node_1.Node) {
                node.toModule(context, out);
            }
            else {
                out.add(JSON.stringify(node));
            }
            if (i < length) {
                out.add(',\n');
            }
        });
        context.indent--;
        pre = context.pre;
        out.add(`\n${pre}])`);
        return out;
    }
}
exports.List = List;
List.prototype.type = 'List';
const list = (value, location) => new List(value, location);
exports.list = list;
